File: scripts/contacts.py
"""
Contact loading and lookup from macOS AddressBook databases.
"""

import os
import re
import sqlite3
from glob import glob


def normalize_phone(phone):
    """Normalize phone number for matching.

    Preserves country code to avoid international number collisions.
    For numbers with 11+ digits starting with country code, keeps the full number.
    For 10-digit numbers (US/Canada without country code), keeps as-is.
    """
    if not phone:
        return ""
    digits = re.sub(r"\D", "", phone)
    if len(digits) == 11 and digits.startswith("1"):
        # US/Canada with country code - normalize to 10 digits
        return digits[1:]
    if len(digits) == 10:
        # US/Canada without country code
        return digits
    # International or other formats - keep full digits to avoid collisions
    return digits


def load_contacts(contacts_dir):
    """Load contact name and ID mappings from AddressBook databases.

    Returns mappings from phone/email to (name, contact_id) tuples.
    The contact_id is used to correctly group identifiers belonging to the
    same contact, avoiding incorrect merges when different people share names.
    """
    phone_to_contact = {}  # normalized_phone -> (name, contact_id)
    email_to_contact = {}  # email -> (name, contact_id)

    db_files = glob(os.path.join(contacts_dir, "*/AddressBook-v22.abcddb"))
    if not db_files:
        print(f"Warning: No AddressBook databases found in {contacts_dir}")
        return phone_to_contact, email_to_contact

    for db_path in db_files:
        try:
            uri = f"file:{db_path}?mode=ro"
            conn = sqlite3.connect(uri, uri=True)
            cursor = conn.cursor()

            # Phone numbers - include Z_PK as unique contact identifier
            cursor.execute("""
                SELECT r.Z_PK, r.ZFIRSTNAME, r.ZLASTNAME, r.ZORGANIZATION, r.ZNICKNAME, p.ZFULLNUMBER
                FROM ZABCDRECORD r
                JOIN ZABCDPHONENUMBER p ON p.ZOWNER = r.Z_PK
                WHERE p.ZFULLNUMBER IS NOT NULL
            """)
            for contact_id, first, last, org, nick, phone in cursor.fetchall():
                name = " ".join(p for p in [first, last] if p) or nick or org
                if name:
                    normalized = normalize_phone(phone)
                    if normalized and normalized not in phone_to_contact:
                        phone_to_contact[normalized] = (name, f"{db_path}:{contact_id}")

            # Emails - include Z_PK as unique contact identifier
            cursor.execute("""
                SELECT r.Z_PK, r.ZFIRSTNAME, r.ZLASTNAME, r.ZORGANIZATION, r.ZNICKNAME, e.ZADDRESS
                FROM ZABCDRECORD r
                JOIN ZABCDEMAILADDRESS e ON e.ZOWNER = r.Z_PK
                WHERE e.ZADDRESS IS NOT NULL
            """)
            for contact_id, first, last, org, nick, email in cursor.fetchall():
                name = " ".join(p for p in [first, last] if p) or nick or org
                if name and email.lower() not in email_to_contact:
                    email_to_contact[email.lower()] = (name, f"{db_path}:{contact_id}")

            conn.close()
        except Exception as e:
            print(f"Warning: Could not read {db_path}: {e}")

    return phone_to_contact, email_to_contact


def lookup_contact(identifier, phone_to_contact, email_to_contact):
    """Look up contact info (name, contact_id) from phone or email.

    Returns (name, contact_id) tuple or (None, None) if not found.
    """
    if not identifier:
        return None, None
    if "@" in identifier:
        return email_to_contact.get(identifier.lower(), (None, None))
    return phone_to_contact.get(normalize_phone(identifier), (None, None))

File: src/exporter/contacts.ts
import { findContactsDatabases, openContactsDb } from './database';
import { normalizePhone } from './phone-utils';
import { ContactInfo } from './types';

/**
 * Contact lookup maps
 */
export interface ContactMaps {
  phoneToContact: Map<string, ContactInfo>;
  emailToContact: Map<string, ContactInfo>;
}

/**
 * Load contact name and ID mappings from AddressBook databases.
 *
 * Returns mappings from phone/email to (name, contactId) tuples.
 * The contactId is used to correctly group identifiers belonging to the
 * same contact, avoiding incorrect merges when different people share names.
 */
export function loadContacts(): ContactMaps {
  const phoneToContact = new Map<string, ContactInfo>();
  const emailToContact = new Map<string, ContactInfo>();

  const dbFiles = findContactsDatabases();

  if (dbFiles.length === 0) {
    console.warn('No AddressBook databases found');
    return { phoneToContact, emailToContact };
  }

  for (const dbPath of dbFiles) {
    try {
      const db = openContactsDb(dbPath);

      // Load phone numbers
      const phoneStmt = db.prepare(`
        SELECT r.Z_PK, r.ZFIRSTNAME, r.ZLASTNAME, r.ZORGANIZATION, r.ZNICKNAME, p.ZFULLNUMBER
        FROM ZABCDRECORD r
        JOIN ZABCDPHONENUMBER p ON p.ZOWNER = r.Z_PK
        WHERE p.ZFULLNUMBER IS NOT NULL
      `);

      for (const row of phoneStmt.iterate() as IterableIterator<{
        Z_PK: number;
        ZFIRSTNAME: string | null;
        ZLASTNAME: string | null;
        ZORGANIZATION: string | null;
        ZNICKNAME: string | null;
        ZFULLNUMBER: string;
      }>) {
        const name = buildContactName(row.ZFIRSTNAME, row.ZLASTNAME, row.ZORGANIZATION, row.ZNICKNAME);
        if (!name) continue;

        const normalized = normalizePhone(row.ZFULLNUMBER);
        if (normalized && !phoneToContact.has(normalized)) {
          phoneToContact.set(normalized, {
            name,
            contactId: `${dbPath}:${row.Z_PK}`,
          });
        }
      }

      // Load emails
      const emailStmt = db.prepare(`
        SELECT r.Z_PK, r.ZFIRSTNAME, r.ZLASTNAME, r.ZORGANIZATION, r.ZNICKNAME, e.ZADDRESS
        FROM ZABCDRECORD r
        JOIN ZABCDEMAILADDRESS e ON e.ZOWNER = r.Z_PK
        WHERE e.ZADDRESS IS NOT NULL
      `);

      for (const row of emailStmt.iterate() as IterableIterator<{
        Z_PK: number;
        ZFIRSTNAME: string | null;
        ZLASTNAME: string | null;
        ZORGANIZATION: string | null;
        ZNICKNAME: string | null;
        ZADDRESS: string;
      }>) {
        const name = buildContactName(row.ZFIRSTNAME, row.ZLASTNAME, row.ZORGANIZATION, row.ZNICKNAME);
        if (!name) continue;

        const email = row.ZADDRESS.toLowerCase();
        if (!emailToContact.has(email)) {
          emailToContact.set(email, {
            name,
            contactId: `${dbPath}:${row.Z_PK}`,
          });
        }
      }

      db.close();
    } catch (error) {
      console.warn(`Could not read ${dbPath}:`, error);
    }
  }

  return { phoneToContact, emailToContact };
}

/**
 * Build contact name from available fields
 */
function buildContactName(
  firstName: string | null,
  lastName: string | null,
  organization: string | null,
  nickname: string | null
): string | null {
  const parts = [firstName, lastName].filter(Boolean);
  if (parts.length > 0) {
    return parts.join(' ');
  }
  return nickname || organization || null;
}

/**
 * Look up contact info from phone or email identifier
 */
export function lookupContact(
  identifier: string,
  contactMaps: ContactMaps
): ContactInfo | null {
  if (!identifier) return null;

  if (identifier.includes('@')) {
    return contactMaps.emailToContact.get(identifier.toLowerCase()) ?? null;
  }

  return contactMaps.phoneToContact.get(normalizePhone(identifier)) ?? null;
}
